package_managers/pypi/fetcher.py

```diff
@@ -0,0 +1,213 @@
+                    # Package names are the href values
+                    self.packages.append(value)
+
+
+class PyPIFetcher(Fetcher):
+    """Custom fetcher for PyPI that handles both simple index and JSON API."""
+    
+    def __init__(self, name: str, config: Config):
+        super().__init__(name, config)
+        self.session = requests.Session()
+        self.base_url = "https://pypi.org"
+        self.rate_limit_delay = 0.1  # seconds between requests
+        self.batch_size = 100  # packages per batch
+        self.max_workers = multiprocessing.cpu_count() * 4  # Number of threads for parallel downloads
+        self.data_dir = "/data/pypi" # We will mount to ./data/pypi on the host
+        self.process_file = os.path.join(self.data_dir, "progress.json") # to store the progress of the fetch
+        self.packages_file = os.path.join(self.data_dir, "packages.txt") # a list of all package names
+        
+    def _save_process(self, batch_num: int, downloaded: int, fetched: int, total: int):
+        """Save current process to progress.json"""
+        process = {
+            "batch_num": batch_num,
+            "downloaded": downloaded,  # successful downloads
+            "fetched": fetched,      # total attempted downloads
+            "total": total,
+            "timestamp": datetime.now().isoformat()
+        }
+        with open(self.process_file, 'w') as f:
+            json.dump(process, f, indent=2)
+    
+    def _load_process(self) -> tuple[int, int, int, int]:
+        """Load process from progress.json if exists"""
+        try:
+            with open(self.process_file, 'r') as f:
+                process = json.load(f)
+                return (
+                    process["batch_num"],
+                    process.get("downloaded", 0),
+                    process.get("fetched", 0),
+                    process["total"]
+                )
+        except (FileNotFoundError, json.JSONDecodeError):
+            return 0, 0, 0, 0
+    
+    def _save_package_list(self, packages: List[str]):
+        """Save package list to packages.txt"""
+        with open(self.packages_file, 'w') as f:
+            for package in packages:
+                f.write(f"{package}\n")
+    
+    def _load_package_list(self) -> List[str]:
+        """Load package list from packages.txt if exists"""
+        try:
+            with open(self.packages_file, 'r') as f:
+                return [line.strip() for line in f if line.strip()]
+        except FileNotFoundError:
+            return []
+
+    def _get_package_list(self) -> List[str]:
+        """Fetch list of all packages from PyPI simple index."""
+        url = f"{self.base_url}/simple/" # PyPI simple index
+        response = self.session.get(url)
+        response.raise_for_status()
+        
+        parser = SimpleIndexParser()
+        parser.feed(response.text)
+        return parser.packages
+    
+    def _get_package_data(self, package_name: str) -> Optional[dict]:
```

Comment:
The new style for something like `Optional[dict]` is `dict | None`.  also I think dict[str, typing.Any] is better.